File: programmers/실패율.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

bool compare(pair<int,float> a,pair<int,float> b){
    if(a.second==b.second)
        return a.first<b.first;
    return a.second>b.second;
}

vector<int> solution(int N, vector<int> stages) {
    vector<int> answer;
    unordered_map<int,float> un;
    float user=stages.size();
    for(int stage=1;stage<=N;++stage){
        float failure=0;
        for(int i=0;i<stages.size();++i){
            if(stage==stages[i])failure++;
        }
        if(user==0){
            if(stage<=N) un.insert(make_pair(stage,0));
        }
        else{
            float ave=(failure/user);
            un.insert(make_pair(stage,ave));
            user-=failure;
        }
    }
    vector<pair<int,float>> elms(un.begin(),un.end());
    sort(elms.begin(),elms.end(),compare);
    for(int i=0;i<elms.size();++i){
        answer.push_back(elms[i].first);
    }
    return answer;
}

File: programmers/Level 1/실패율.js
function solution(N, stages) {
    let answer= [];
    let total = stages.length;
    for(let i=1;i<=N+1;++i){
        //filter를 통해 i 스테이지에 실패한 사람의 수를 user에 저장
        let users = stages.filter(n=>n===i).length
        //[스테이지, 실패율] 을 answer에 할당
        answer.push([i, users/total]);
        //총 인원의 수를 실패한 사람의 수만큼 minus
        total-=users;
    }
    //for문의 루프를 N+1까지 해두었기 때문에 마지막 N+1 스테이지를 pop()
    answer.pop();
    //두번째 원소를 기준으로 내림차순 정렬
    answer= answer.sort((a,b)=>b[1]-a[1]);
    return answer.map(a=>a[0]);
}
/*
N = 5
2,1,2,6,2,4,3,3
stages   failure 
 1     : 1 / 8  0.125
 2     : 3 / 7  0.428  
 3     : 2 / 4  0.5
 4     : 1 / 2  0.5
 5     : 0 / 1  0
 => 3,4,2,1,5
 6 : 8 / 1
 N + 1 은 마지막 스테이지까지 클리어 한 사용자
 */
